### DISCUSSION TITLE
Subscribing to Codesys standard OPC UA events

### DISCUSSION THREAD

**Discussion by mozesa** [FreeOpcUa/opcua-asyncio#572] [Q&A]
**Hello**,

First of all thanks for your effort putting into this precious project.

I am using [Codesys][1] runtime which provides OPC UA Server interface.
It works fine with `opcua-asyncio`, I have been using `subscribe_data_change` in production for almost half a year without any issue.

Now, I would like to do the same for events with `subscribe_events`.
I created the wanted OPC UA events based on [this guide][2] and I can see those ones using [UaExpert][3] but cannot using the below snippet.

```python
import asyncio

from asyncua import Client


class SubscriptionHandler:
    def event_notification(self, event):
        print(f"{event=}")


async def main():
    client = Client(url="opc.tcp://127.0.0.1:4840/")
    async with client:
        obj = await client.nodes.root.get_child(
            ["0:Objects", "2:DeviceSet", "4:CODESYS Control Win V3 x64", "4:Application"]
        )
        events = await client.nodes.root.get_child(
            ["0:Types", "0:EventTypes", "0:BaseEventType", "0:ConditionType"]
        )

        # We create a Client Subscription.
        handler = SubscriptionHandler()
        sub = await client.create_subscription(500, handler)
        handle = await sub.subscribe_events(obj, events)
        await asyncio.sleep(60)
        await sub.unsubscribe(handle)
        await sub.delete()


if __name__ == "__main__":
    asyncio.run(main())
```

On the below image I marked red the events I would like to subscribe to.
![image](https://user-images.githubusercontent.com/49973426/118613670-4473ac00-b7bf-11eb-9aa0-24b4c8718abc.png)

What I am doing wrong?

I would appreciate your help. 

Thanks in advance.

[1]: https://www.codesys.com/
[2]: https://help.codesys.com/webapp/_cds_runtime_opc_ua_server;product=codesys;version=3.5.17.0#creating-events-in-the-codesys-project
[3]: https://www.unified-automation.com/products/development-tools/uaexpert.html

**Comment by AndreasHeine** [comment]
https://github.com/FreeOpcUa/opcua-asyncio/blob/51fb682aea7f23efb69ef9f0ff3660d8bce1d2ac/asyncua/common/subscription.py#L189

is "4:Application" really the emitting node or is it "Server" ?

**Reply by mozesa** [reply]
```python
import asyncio

from asyncua import Client


class SubscriptionHandler:
    def event_notification(self, event):
        print(f"{event=}")


async def main():
    client = Client(url="opc.tcp://127.0.0.1:4840")
    async with client:
        obj = await client.nodes.root.get_child(
            [
                "0:Objects",
                "0:Server",
                # "4:CODESYS Control Win V3 x64",
                # "4:Application",
            ]
        )
        events = await client.nodes.root.get_child(
            [
                "0:Types",
                "0:EventTypes",
                "0:BaseEventType",
                "0:ConditionType",
            ]
        )

        # We create a Client Subscription.
        handler = SubscriptionHandler()
        sub = await client.create_subscription(500, handler)
        handle = await sub.subscribe_events(obj, events)
        await asyncio.sleep(60)
        await sub.unsubscribe(handle)
        await sub.delete()


if __name__ == "__main__":
    asyncio.run(main())
```

I altered the code, but no success. I don't get any events.
Strange b/c I can see those events in UaExpert (Event View) using Application as object.
It can be that I am wrong (as so far I used only `subscribe_data_change`).

![Codesys project](https://user-images.githubusercontent.com/49973426/118707725-28055d00-b81b-11eb-87c9-1332ec14cf69.png)
![UaExpert](https://user-images.githubusercontent.com/49973426/118707741-2b98e400-b81b-11eb-8a8b-45571a09c8ad.png)
![image](https://user-images.githubusercontent.com/49973426/118707901-61d66380-b81b-11eb-8af7-d3a185600ab6.png)

**Reply by AndreasHeine** [reply]
have you tried the codesys device instead of server? the emitting node must have a reference which shows that it emitts events

**Reply by mozesa** [reply]
No, it doesn't report anything at all. 😞 

```python
import asyncio

from asyncua import Client


class SubscriptionHandler:
    def event_notification(self, event):
        print(f"{event=}")


async def main():
    client = Client(url="opc.tcp://127.0.0.1:4840")
    async with client:
        obj = await client.nodes.root.get_child(
            [
                "0:Objects",
                "2:DeviceSet",
                "4:CODESYS Control Win V3 x64",
                # "4:Application",
            ]
        )
        events = await client.nodes.root.get_child(
            [
                "0:Types",
                "0:EventTypes",
                "0:BaseEventType",
                "0:ConditionType",
            ]
        )

        # We create a Client Subscription.
        handler = SubscriptionHandler()
        sub = await client.create_subscription(500, handler)
        handle = await sub.subscribe_events(obj, events)
        print("*** Sleep ***")
        await asyncio.sleep(60)
        await sub.unsubscribe(handle)
        await sub.delete()


if __name__ == "__main__":
    asyncio.run(main())
```

**Comment by AndreasHeine** [comment]
please post the references of "server", "4:CODESYS Control Win V3 x64" and "4:Application"

ther must be a reference which says "GeneratesEvents" or "HasCondition"!?

**Reply by mozesa** [reply]
I really appreciate your help!

![image](https://user-images.githubusercontent.com/49973426/118720835-635b5800-b82a-11eb-8786-f6afa4c803ca.png)

As I have mentioned, I get the events using UaExpert.
I enclosed some pictures about that too.
![image](https://user-images.githubusercontent.com/49973426/118721576-568b3400-b82b-11eb-88eb-0d8b4e3bf7d5.png)

**Reply by AndreasHeine** [reply]
there it is the "HasNotifier" reference 😉 could you try to subscribe the baseevent on server? this should show all events which are emitted!

**Reply by AndreasHeine** [reply]
https://reference.opcfoundation.org/v104/Core/docs/Part3/7.18/

**Reply by mozesa** [reply]
It doesn't show anything. 😞 

The events are visible only in UaExpert.

```python
import asyncio

from asyncua import Client


class SubscriptionHandler:
    def event_notification(self, event):
        print(f"{event=}")


async def main():
    client = Client(url="opc.tcp://127.0.0.1:4840")
    async with client:
        obj = await client.nodes.root.get_child(
            [
                "0:Objects",
                "0:Server",
                # "4:CODESYS Control Win V3 x64",
                # "4:Application",
            ]
        )
        events = await client.nodes.root.get_child(
            [
                "0:Types",
                "0:EventTypes",
                "0:BaseEventType",
                # "0:ConditionType",
            ]
        )

        # We create a Client Subscription.
        handler = SubscriptionHandler()
        sub = await client.create_subscription(50, handler)
        handle = await sub.subscribe_events(obj, events)
        print("*** Sleep ***")
        await asyncio.sleep(60)
        await sub.unsubscribe(handle)
        await sub.delete()


if __name__ == "__main__":
    asyncio.run(main())
```

![image](https://user-images.githubusercontent.com/49973426/118762755-0639c380-b877-11eb-8521-6c6f21189a56.png)

**Reply by mozesa** [reply]
I creating the subscription with the following parameters with UaExpert.

Creating new subscription:
* ClientHandle=2
* PublishingEnable=1
* LifeTimeCount=2400
* MaxKeepAliveCount=10
* Priority=0
* PublishingInterval=500

I can see from the server information (CurrentSubscriptionCount) that the subscription is successful.

I enclosed the logs, I am trying to find the differences.
[UaExpert_log.txt](https://github.com/FreeOpcUa/opcua-asyncio/files/6506950/UaExpert_log.txt)
[asyncua_log.txt](https://github.com/FreeOpcUa/opcua-asyncio/files/6506951/asyncua_log.txt)

**Comment by mozesa** [comment]
Using [QuickOPC][1] I can subscribe to events.

```python
def subscribe_to_events():
    endpoint_descriptor = UAEndpointDescriptor("opc.tcp://L042IEAMO2:4840")

    # Instantiate the client object and hook events
    client = EasyUAClient()
    client.EventNotification += print

    # client.SubscribeEvent(endpoint_descriptor, AddressSpace.Standard.UAObjectIds.Server, 1000)
    IEasyUAClientExtension.SubscribeEvent(
        client,
        endpoint_descriptor,
        UANodeDescriptor(AddressSpace.Standard.UAObjectIds.Server),
        1_000,
    )
    Threading.Thread.Sleep(30 * 1000)
```

```text
{Isolated=False} [] Success
{Isolated=False} [] Success; Refresh; RefreshInitiated
{Isolated=False} [] Success; Refresh; RefreshComplete
{Isolated=False} [] Success; (10 field results) [Application] 1000! "TRUE" @19/05/2021 08:39:59
{Isolated=False} [] Success; (10 field results) [Application] 1000! "TRUE" @19/05/2021 08:39:59
{Isolated=False} [] Success; (10 field results) [Application] 1000! "TRUE" @19/05/2021 08:39:59
{Isolated=False} [] Success; (10 field results) [Application] 1000! "TRUE" @19/05/2021 08:40:04
{Isolated=False} [] Success; (10 field results) [Application] 1000! "TRUE" @19/05/2021 08:40:04
{Isolated=False} [] Success; (10 field results) [Application] 1000! "TRUE" @19/05/2021 08:40:04
```

So what is the difference between `opcua-asyncio` vs `UaExpert` / `QuickOPC`.


[1]: http://opclabs.doc-that.com/files/onlinedocs/QuickOpc/Latest/User's%20Guide%20and%20Reference-QuickOPC/Subscribing%20to%20OPC%20Events2.html

**Comment by mozesa** [comment]
```python
class SimpleAttributeOperand(FrozenClass):

    def __init__(self):
        self.TypeDefinitionId = NodeId(2041)  # Using this NodeId it works, but why?!
```

Previously it was `TypeDefinitionId:i=0`.

Is it a bug?

**Reply by swamper123** [reply]
2041 is the BaseEventType, which should always work.

**Reply by mozesa** [reply]
```python
async def main():
    client = Client(url="opc.tcp://127.0.0.1:4840/")
    async with client:
        obj = await client.nodes.root.get_child(
            [
                "0:Objects",
                "2:DeviceSet",
                "4:CODESYS Control Win V3 x64",
                "4:Application",
            ]
        )
        events = await client.nodes.root.get_child(
            [
                "0:Types",
                "0:EventTypes",
                "0:BaseEventType",
                "0:ConditionType",
            ]
        )

        # We create a Client Subscription.
        handler = SubscriptionHandler()

        params = CreateSubscriptionParameters()
        params.RequestedPublishingInterval = 500
        params.RequestedLifetimeCount = 2400
        params.RequestedMaxKeepAliveCount = 10
        params.MaxNotificationsPerPublish = 10_000
        params.PublishingEnabled = 1
        params.Priority = 0

        sub = await client.create_subscription(params, handler)
        handle = await sub.subscribe_events(obj, events, queuesize=50)
        print("*** Sleep ***")
        await asyncio.sleep(60)
        await sub.unsubscribe(handle)
        await sub.delete()
```

`events` == 2782

![image](https://user-images.githubusercontent.com/49973426/118794817-3e063280-b89a-11eb-8d32-4849ab097c87.png)

I don't get any events.

But when I manually overwrite it in `/asyncua/ua/uaprotocol_auto.py` with `self.TypeDefinitionId = NodeId(2782)`, I finally get all the events.

![image](https://user-images.githubusercontent.com/49973426/118795231-ace38b80-b89a-11eb-8f4f-bef8e40e440d.png)


Do I miss something?

Strange. The line which is responsible for `TypeDefinitionId` is commented out.

```python
async def where_clause_from_evtype(evtypes):
    cf = ua.ContentFilter()
    el = ua.ContentFilterElement()
    # operands can be ElementOperand, LiteralOperand, AttributeOperand, SimpleAttribute
    # Create a clause where the generate event type property EventType
    # must be a subtype of events in evtypes argument

    # the first operand is the attribute event type
    op = ua.SimpleAttributeOperand()
    # op.TypeDefinitionId = evtype.nodeid
```

![image](https://user-images.githubusercontent.com/49973426/118796344-d2bd6000-b89b-11eb-8e4c-49df72e52fb2.png)

It never gets the set to the proper `nodeid`.

**Reply by swamper123** [reply]
2782 is the index for the ConditionType.
https://raw.githubusercontent.com/FreeOpcUa/opcua-asyncio/master/asyncua/ua/object_ids.py

Just a guess from my side: Could you subscribe to a single EventType? Like just the BaseEventType? 
And then try to create multiple single handles for each event Type?

**Reply by mozesa** [reply]
The problem that `SimpleAttributeOperand` is a `FrozenClass` subtype so it never gets updated later, this is the reason for it has `i=0` id.
Plus it is commented out too 😄

**Comment by mozesa** [comment]
I am at the latest master commit `rev = "51fb682"`.

![image](https://user-images.githubusercontent.com/49973426/118803102-3ac37480-b8a3-11eb-8ef3-9cfffd478353.png)

The problem that `evfilter` `TypeDefinitionId` is wrong, I mean the `Identifier` == 0.

Or am I wrong?

**Reply by mozesa** [reply]
![image](https://user-images.githubusercontent.com/49973426/118805304-11581800-b8a6-11eb-8d95-dc6cc13e6f35.png)

Explicitly giving Node's Identifier, works.
I know it is a dirty workaround.

Btw, does this function work properly or it is a bug?

**Comment by AndreasHeine** [comment]
i am on 0.9.14 and i tested it against the prosys simserver:

![grafik](https://user-images.githubusercontent.com/56362817/119875360-f715de00-bf26-11eb-8e76-ef4effbc849a.png)

**Reply by mozesa** [reply]
Thanks for your help.

I have been using the aforementioned `NodeId(2782)` workaround (via monkey patching that specific class).

It works properly for Codesys OPCUA server.

I understand that it works for for instance prosys server without any issues. But it seems to me that it doesn't work for Codesys OPCUA servers. I compared the EventFilter clause (UaExpert vs FreeOPCUA) and the only difference was/is the EventFiler`s SelectClause`s NodeId == 0.

**Reply by AndreasHeine** [reply]
i would guess 0 is just a default which never gots overwritten in the code... but i need some time to look deeper unfortunatly i have no codesys stuff to test against...

**Reply by mozesa** [reply]
You can download Codesys [developer](https://store.codesys.com/codesys.html).

After installing it, it will also install a soft plc (Runtime Win) and a gateway, all in all, you can freely use it to test against freeopcua (it is a full version for 2 hours).

Anyway I can help you as much as you would like. I really want to release the production version of my program without that nightmarish monkey patch.

**Reply by AndreasHeine** [reply]
can you provide a test project?

**Reply by mozesa** [reply]
Here you are the [project](https://we.tl/t-L7D4WHu5gc).

If you have any questions, let me know. 

Thanks for your help.

**Reply by ghost** [reply]
> i would guess 0 is just a default which never gots overwritten in the code... but i need some time to look deeper unfortunatly i have no codesys stuff to test against...

That's correct,  
in our fork of this library [asyncua/common/events.py](https://github.com/KUZ-Leipzig/opcua-asyncio/blob/master/asyncua/common/events.py) I've set the `TypeDefinitionId` with the `NodeId` of `evtypes`, which was subscribed to and the subscription is working for the Euromap77/83 server of our machine.
But it's the first try, so it can potentially cause many bugs for other servers or subscripions of other events.

**Comment by mozesa** [comment]
@AndreasHeine I really appreciate your help!

I recreated the `asyncua` part based on your snippet, but unfortunately it did not help.

```python
import asyncio

from asyncua import Client


class Handler:
    def event_notification(self, event) -> None:
        print(f"{event=}")


async def async_main():
    client = Client("opc.tcp://L042IEAMO2:4840")
    async with client:
        ev_src = client.get_node("ns=4;s=|var|CODESYS Control Win V3 x64.Application")
        ev_type_node = client.get_node("ns=0;i=2782")
        handler = Handler()
        sub = await client.create_subscription(500, handler)
        await sub.subscribe_events(ev_src, ev_type_node)
        await asyncio.sleep(60)


if __name__ == "__main__":
    asyncio.run(async_main())
```

I could narrow this issue down to the following facts (please see the screenshot).
* (1) `ev_src` node.
* (1.1) `ev_type_node` node.
* (2) The `SelectClauses` based on UaExpert's log.
You can clearly see that each element has `TypeDefinitionId` with value `NS0|Numeric|2041` .
* (3) The same `SelectClauses` resulted by `asyncua`.
You can see that all the `TypeDefinitionId`s value is `i=0`.

![image](https://user-images.githubusercontent.com/49973426/119947460-0da84d80-bf98-11eb-9745-a18745b8c2fb.png)

I don't know why but Codesys OPC UA server does require the proper `TypeDefinitionId` set.
I have done some tests using [QuickOPC](https://www.opclabs.com/products/quickopc) which also works fine with Codesys OPC UA server.

In the [documentation](https://reference.opcfoundation.org/v104/Core/docs/Part4/7.4.4/#7.4.4.5) it is written that 

> typeDefinitionId is NodeId of a TypeDefinitionNode.

which implies me that it really should be `i=2782`, I thought.


After monkey patching `SimpleAttributeOperand` class
------
Asyncua works fine.
Notice that now `TypeDefinitionId` is `2782`.

![image](https://user-images.githubusercontent.com/49973426/119949040-c15e0d00-bf99-11eb-9d49-56e415ae946e.png)
![image](https://user-images.githubusercontent.com/49973426/119949146-dfc40880-bf99-11eb-98d6-aec5cb8d7b93.png)

Thanks for your support.

**Reply by Saleem344** [reply]
Hi I'm Unale to read alarms and events from plc 
out put:
Revised values returned differ from subscription values: CreateSubscriptionResult(SubscriptionId=2272068970, RevisedPublishingInterval=500.0, RevisedLifetimeCount=7200, RevisedMaxKeepAliveCount=2400)

```
async def main():
    url = "opc.tcp://192.0.1.212:4840"
    async with Client(url=url) as client:
        obj = client.get_node("ns=0;i=2253")
        myevent = client.get_node("i=0;i=2782")
        msclt = SubHandler()
        sub = await client.create_subscription(500, msclt)
        await sub.subscribe_events(obj, myevent)
        await asyncio.sleep(30)
        # await sub.unsubscribe(handle)
        # await sub.delete()


if __name__ == "__main__":
    asyncio.run(main())
```
OPC UA server is Bachmann plc using UaExpert I can all alarms and event and acknowledge them but not woring with asyncua.


![Screenshot from 2021-09-16 12-41-44](https://user-images.githubusercontent.com/61460632/133566870-a16c9251-383d-43a5-b13b-3c63203dd113.png)
![Screenshot from 2021-09-16 12-41-54](https://user-images.githubusercontent.com/61460632/133566881-9c125aeb-e63c-455c-9f12-a87b46d5e607.png)

Can any one help me?

Thanks you.

**Reply by AndreasHeine** [reply]
Rrevised values: thats normal you requested parameters which the server is not capable of and the server send you new parameters to use!

Ack Alarms: how do you ack them? Codesnipped?

**Reply by Saleem344** [reply]
> Rrevised values: thats normal you requested parameters which the server is not capable of and the server send you new parameters to use!
> 
> Ack Alarms: how do you ack them? Codesnipped?

Hi @AndreasHeine 
Thanks for your response.
I would like to know how to read the events from python client because I can see it in UaExpert.

ack I'm able to do it from UaExpert not from python.

Thanks.

**Reply by AndreasHeine** [reply]
> > Rrevised values: thats normal you requested parameters which the server is not capable of and the server send you new parameters to use!
> > 
> > Ack Alarms: how do you ack them? Codesnipped?
> 
> Hi @AndreasHeine 
> Thanks for your response.
> I would like to know how to read the events from python client because I can see it in UaExpert.
> 
> ack I'm able to do it from UaExpert not from python.
> 
> Thanks.
> 

https://github.com/FreeOpcUa/opcua-asyncio/blob/master/examples/client_to_uanet_alarm_conditions.py

**Reply by Saleem344** [reply]
> > > Rrevised values: thats normal you requested parameters which the server is not capable of and the server send you new parameters to use!
> > > Ack Alarms: how do you ack them? Codesnipped?
> > 
> > 
> > Hi @AndreasHeine
> > Thanks for your response.
> > I would like to know how to read the events from python client because I can see it in UaExpert.
> > ack I'm able to do it from UaExpert not from python.
> > Thanks.
> 
> https://github.com/FreeOpcUa/opcua-asyncio/blob/master/examples/client_to_uanet_alarm_conditions.py

Hi @AndreasHeine 
It is giving some other output not the required event data.

Output:

 New event received: %r Event(['AudibleEnabled:None', 'FirstInGroupFlag:None', 'InputNode:None', 'MaxTimeShelved:None', 'OffDelay:None', 'OnDelay:None', 'ReAlarmRepeatCount:None', 'ReAlarmTime:None', 'SuppressedOrShelved:None', 'BranchId:None', 'ClientUserId:None', 'ConditionClassId:None', 'ConditionClassName:None', 'ConditionName:None', 'ConditionSubClassId:None', 'ConditionSubClassName:None', 'Retain:None', 'EventId:None', 'EventType:None', 'LocalTime:None', 'Message:None', 'ReceiveTime:None', 'Severity:None', 'SourceName:None', 'SourceNode:None', 'Time:None', 'ActiveState:None', 'ActiveState/EffectiveDisplayName:None', 'ActiveState/EffectiveTransitionTime:None', 'ActiveState/FalseState:None', 'ActiveState/Id:None', 'ActiveState/TransitionTime:None', 'ActiveState/TrueState:None', 'AudibleSound:None', 'EnabledState:None', 'EnabledState/Id:None', 'LatchedState:None', 'LatchedState/FalseState:None', 'LatchedState/Id:None', 'LatchedState/TransitionTime:None', 'LatchedState/TrueState:None', 'OutOfServiceState:None', 'OutOfServiceState/FalseState:None', 'OutOfServiceState/Id:None', 'OutOfServiceState/TransitionTime:None', 'OutOfServiceState/TrueState:None', 'SilenceState:None', 'SilenceState/FalseState:None', 'SilenceState/Id:None', 'SilenceState/TransitionTime:None', 'SilenceState/TrueState:None', 'SuppressedState:None', 'SuppressedState/FalseState:None', 'SuppressedState/Id:None', 'SuppressedState/TransitionTime:None', 'SuppressedState/TrueState:None', 'AckedState:None', 'AckedState/FalseState:None', 'AckedState/Id:None', 'AckedState/TransitionTime:None', 'AckedState/TrueState:None', 'ConfirmedState:None', 'ConfirmedState/FalseState:None', 'ConfirmedState/Id:None', 'ConfirmedState/TransitionTime:None', 'ConfirmedState/TrueState:None', 'Comment:None', 'Comment/SourceTimestamp:None', 'EnabledState/EffectiveDisplayName:None', 'EnabledState/EffectiveTransitionTime:None', 'EnabledState/FalseState:None', 'EnabledState/TransitionTime:None', 'EnabledState/TrueState:None', 'LastSeverity:None', 'LastSeverity/SourceTimestamp:None', 'Quality:None', 'Quality/SourceTimestamp:None', 'NodeId:None'])

**Reply by MadTheT** [reply]
Hi Saleem344, 
I have the same problem with subscribing to events as you that no data a send by the server (Wireshark debug shows that there are really no data send by the server). I use an  ARSim Server by B&R. I can provide the wireshark logs if necessary. My guess is that im doing something wrong by the connection establishing or in the subscription settings. 
I tried with the RnCommtest (https://github.com/br-automation-com/RnCommTest-Windows) which works fine. 

Have you solved your problem?

Thanks

**Comment by danmoser** [comment]
Folks, FYI a recent issue I faced with a TwinCAT server is similar to the problem described here: https://github.com/FreeOpcUa/opcua-asyncio/discussions/1205
